connaisseur/validators/cosign/cosign_validator.py

```diff
@@ -64,48 +64,46 @@ async def validate(
             vals=self.vals, threshold=threshold, required=required
         )
 
-    def __get_pinned_keys(self, key_name: str = None, required: list = None):
+    def __get_pinned_keys(
+        self, key_name: str = None, required: list = None, threshold: int = None
+    ):
         """
         Extract the pinned key(s) selected for validation from the list of trust roots.
         """
         key_name = key_name or "default"
+        available_keys = list(map(lambda k: k["name"], self.trust_roots))
 
+        # generate list of pinned keys
         if key_name == "*":
-            keys = {
-                k["name"]: {
-                    "name": k["name"],
-                    "key": Key("".join(k["key"])),
-                    "digest": None,
-                    "error": None,
-                }
-                for k in self.trust_roots
-            }
-
-            # check if all `required` are present in trust roots
-            not_found = set(required) - {k["name"] for k in self.trust_roots}
-            if not_found:
-                msg = 'Trust roots "{key_names}" not configured for validator "{validator_name}".'
-                raise NotFoundException(
-                    message=msg,
-                    key_names=", ".join(not_found),
-                    validator_name=self.name,
-                )
+            if len(required) >= threshold:
```

Comment:
1. How can the threshold be lower than the number of required signatures?
2. `len` and `>=` doesn't work if either side is `None`.

Reply:
1. as implemented in the `validate` method (you also made a proposal for that), I chose for implementation that `threshold` is 1 if `required` contains a list of required trust roots and there is no direct specification of `threshold`. that means if you only provide a setting for `required` in `helm/values.yaml`, threshold is set to 1, however, connaisseur will enforce those 2 signatures. Here, it is used to shorten the required number of validations if the validator contains a lot of trust roots but only a few are actually required and no other threshold is used. I think I also had an actual reason which goes beyond throwing the exception in __apply_policy, but cannot recall rn
2. based on `validate` method neither can be `None`, but I agree that this is misleading with the defaults. Would you advise to drop the defaults?

Reply:
2. Yes!

Reply:
thx for advice. changed

Reply:
in a stupid way, but now it works :stuck_out_tongue_winking_eye: